index: move disk read lookups to base class

furszy opened the pull request:
Combining common refactors from #24230 and #26966, aiming to move both efforts forward while reducing their size and review burden.

Broadly, #24230 focuses on enabling indexes to run in a separate process, and #26966 aims to parallelize the indexes initial synchronization process. A shared prerequisite for both is ensuring that only the base index class interacts with the node’s chain internals - child index classes should instead operate solely through chain events.

This PR moves disk read lookups from child index classes to the base index class. It also includes a few documentation improvements and a test-only code cleanup.

ryanofsky committed:
indexes, refactor: Remove remaining CBlockIndex* uses in index Rewind methods

Move ReadBlock code from CoinStatsIndex::CustomRewind to BaseIndex::Rewind

Move ReadUndo code from CoinStatsIndex::ReverseBlock to BaseIndex::Rewind

This commit does change behavior slightly. Since the new CustomRemove
methods only take a single block at a time instead of a range of
disconnected blocks, when they call CopyHeightIndexToHashIndex they will
now do an index seek for each removed block instead of only seeking once
to the height of the earliest removed block. Seeking instead of scanning
is a little worse for performance if there is a >1 block reorg, but
probably not noticeable unless the reorg is very deep.

ryanofsky committed:
indexes, refactor: Stop requiring CBlockIndex type to call IsBIP30Unspendable

This commit does not change behavior in any way.

furszy committed:
index: remove CBlockIndex access from CustomAppend()

Moved CBlockUndo disk read lookups from child index classes to
the base index class.

The goal is for child index classes to synchronize only through
events, without directly accessing the chain database.

This change will enable future parallel synchronization mechanisms,
reduce database access (when batched), and contribute toward the
goal of running indexes in a separate process (with no chain
database access).

Besides that, this commit also documents how NextSyncBlock() behaves.
It is not immediately clear this function could return the first
block after the fork point during a reorg.

furszy committed:
test: indexes, avoid creating threads when sync runs synchronously

The indexes test call StartBackgroundSync(), which spawns a thread to run Sync(),
only for the test thread to wait for it to complete by calling IndexWaitSynced().

So, since the sync is performed synchronously, we can skip the extra thread creation
entirely and call Sync() directly.

ryanofsky reviewed (approved):
Code review ACK 3f390efa313f4ac43e9134d5299d7faec74cca60. Nice changes! This gets rid of blockstorage and CBlockIndex* accesses from index subclasses during sync to decouple the indexes from node internals and make new indexes more straightforward to write in the future. It should also simplify the parallelization PR and avoid the need for refactoring there.

sedited reviewed (commented):
Approach ACK

This all felt very familiar :)

Can you double-check the commit message of 50696f07784b5dcf04a398f7592dead4dde6d2e9. I think there is a word too much in the first and second sentence.

TheCharlatan commented on src/index/base.cpp:
```diff
@@ -153,6 +153,39 @@ static const CBlockIndex* NextSyncBlock(const CBlockIndex* pindex_prev, CChain&
+{
+    interfaces::BlockInfo block_info = kernel::MakeBlockInfo(pindex, block_data);
+
+    CBlock block;
```
Of course, not sure what tripped me up here :(

sedited reviewed (approved):
ACK 4c5ec3a70a47798d8451da65c9c0059955cde937

mzumsande reviewed (commented):
Concept ACK

furszy commented on src/index/blockfilterindex.cpp:
```diff
@@ -316,15 +316,15 @@ bool BlockFilterIndex::Write(const BlockFilter& filter, uint32_t block_height, c
     // height index to the hash index so we can still find them when the height index entries are
     // overwritten.
-    if (!CopyHeightIndexToHashIndex(*db_it, batch, m_name, new_tip.height, current_tip.height)) {
+    if (!CopyHeightIndexToHashIndex(*db_it, batch, m_name, block.height - 1, block.height)) {
```
This reminds me to my own [comment](https://github.com/bitcoin/bitcoin/pull/24230/commits/707ff84981d7f61d467093cb0aec7eee55276c52#r1307972744) hehe. Will do it here.

furszy committed:
refactor: index, simplify CopyHeightIndexToHashIndex to process single block

The function previously handled a range of heights but now copies only a
single entry. This change simplifies the code to match its current usage.

sedited reviewed (approved):
Re-ACK 029ba1a21d570f7db6c4366ec9a30a381b56d6fb

davidgumberg commented on src/interfaces/chain.h:
```diff
@@ -327,6 +327,15 @@ class Chain
         virtual void chainStateFlushed(ChainstateRole role, const CBlockLocator& locator) {}
     };
 
+    //! Options specifying which chain notifications are required.
```
https://github.com/bitcoin/bitcoin/pull/32694/commits/6f1392cc42cde638773f2b697d7d2c58abcdc860 (_indexes, refactor: Remove remaining CBlockIndex* uses in index Rewind methods_)

---

nit:

```suggestion
    //! Options specifying data required for chain notifications.
```

davidgumberg commented on src/interfaces/chain.h:
```diff
@@ -327,6 +327,15 @@ class Chain
+    struct NotifyOptions
+    {
+        //! Include block data with block disconnected notifications.
+        bool disconnect_data = false;
```
https://github.com/bitcoin/bitcoin/pull/32694/commits/6f1392cc42cde638773f2b697d7d2c58abcdc860 (_indexes, refactor: Remove remaining CBlockIndex* uses in index Rewind methods_)

---

feel-free-to-disregard-style-nit:

I feel this would be  more consistent as:

```suggestion
        bool disconnect_block_data = false;
```

davidgumberg reviewed (approved):
ACK 029ba1a21d570f7db6c

Left some nits, and general observations that don't need to be addressed in this PR or at all.

davidgumberg commented on src/index/blockfilterindex.cpp:
```diff
@@ -289,30 +289,25 @@ bool BlockFilterIndex::Write(const BlockFilter& filter, uint32_t block_height, c
 {
-    DBHeightKey key(start_height);
+    DBHeightKey key(height);
     db_it.Seek(key);
```
The `GetKey()` call and `key.height` check below protects against the situation where an entry with a height greater than the key we're seeking exists, but the height we're seeking doesn't exist.  In the instance where the height we're seeking is greater than any entry which exists, which should never happen,`db_it.GetKey()` will assert and crash.

Just to be clear though, I think this is out-of-scope and the PR is ready to merge as-is.

mzumsande reviewed (commented):
Code Review ACK 029ba1a21d570f7db6c4366ec9a30a381b56d6fb

achow101 commented:
ACK 029ba1a21d570f7db6c4366ec9a30a381b56d6fb
